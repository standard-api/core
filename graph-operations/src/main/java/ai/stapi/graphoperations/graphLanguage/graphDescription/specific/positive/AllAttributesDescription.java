package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.GraphBaseTypes;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;
import java.util.List;

public class AllAttributesDescription extends AbstractPositiveGraphDescription {

  public static final String SERIALIZATION_TYPE = "AllAttributesDescription";

  private AllAttributesDescriptionParameters allAttributesDescriptionParameters;

  protected AllAttributesDescription() {
  }

  public AllAttributesDescription(List<GraphDescription> childDescriptions) {
    super(AllAttributesDescription.SERIALIZATION_TYPE, GraphBaseTypes.VALUE_TYPE,
        childDescriptions);
  }

  public AllAttributesDescription(GraphDescription... childDescriptions) {
    super(AllAttributesDescription.SERIALIZATION_TYPE, GraphBaseTypes.VALUE_TYPE,
        childDescriptions);
  }

  @Override
  public GraphDescriptionParameters getParameters() {
    return this.allAttributesDescriptionParameters;
  }
}
