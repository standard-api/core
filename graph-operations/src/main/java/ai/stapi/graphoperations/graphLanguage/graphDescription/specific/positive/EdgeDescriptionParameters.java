package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class EdgeDescriptionParameters implements GraphDescriptionParameters {

  private String edgeType;

  private EdgeDescriptionParameters() {
  }

  public EdgeDescriptionParameters(String edgeType) {
    this.edgeType = edgeType;
  }

  public String getEdgeType() {
    return edgeType;
  }
}
