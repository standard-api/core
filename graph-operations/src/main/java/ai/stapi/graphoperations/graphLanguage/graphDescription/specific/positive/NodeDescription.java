package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.ArrayList;
import java.util.List;

public class NodeDescription extends AbstractNodeDescription {

  public static final String SERIALIZATION_TYPE = "NodeDescription";

  protected NodeDescription() {

  }

  public NodeDescription(
      NodeDescriptionParameters parameters,
      List<GraphDescription> childDeclarations
  ) {
    super(SERIALIZATION_TYPE, childDeclarations);
    this.parameters = parameters;
  }

  public NodeDescription(NodeDescriptionParameters parameters) {
    super(SERIALIZATION_TYPE, new ArrayList<>());
    this.parameters = parameters;
  }

  public NodeDescription(
      NodeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
  }

  public NodeDescription(
      String nodeType,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, new NodeDescriptionParameters(nodeType), childDeclarations);
  }
}
