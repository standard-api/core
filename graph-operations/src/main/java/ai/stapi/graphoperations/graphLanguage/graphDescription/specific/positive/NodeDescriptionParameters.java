package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class NodeDescriptionParameters implements GraphDescriptionParameters {

  private String nodeType;

  private NodeDescriptionParameters() {

  }

  public NodeDescriptionParameters(String nodeType) {
    this.nodeType = nodeType;
  }

  public String getNodeType() {
    return nodeType;
  }
}