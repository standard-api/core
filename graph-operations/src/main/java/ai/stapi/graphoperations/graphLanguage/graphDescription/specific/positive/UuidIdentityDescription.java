package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.GraphBaseTypes;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;
import java.util.ArrayList;
import java.util.List;

public class UuidIdentityDescription extends AbstractPositiveGraphDescription {

  public static final String SERIALIZATION_TYPE = "UuidIdentityDescription";
  private UuidIdentityDescriptionParameters parameters;

  public UuidIdentityDescription() {
    super(
        SERIALIZATION_TYPE,
        GraphBaseTypes.VALUE_TYPE,
        new ArrayList<>()
    );
    this.parameters = new UuidIdentityDescriptionParameters();
  }

  public UuidIdentityDescription(List<GraphDescription> children) {
    super(
        SERIALIZATION_TYPE,
        GraphBaseTypes.VALUE_TYPE,
        children
    );
    this.parameters = new UuidIdentityDescriptionParameters();
  }

  public UuidIdentityDescription(GraphDescription... children) {
    super(
        SERIALIZATION_TYPE,
        GraphBaseTypes.VALUE_TYPE,
        children
    );
    this.parameters = new UuidIdentityDescriptionParameters();
  }

  @Override
  public GraphDescriptionParameters getParameters() {
    return this.parameters;
  }
}
