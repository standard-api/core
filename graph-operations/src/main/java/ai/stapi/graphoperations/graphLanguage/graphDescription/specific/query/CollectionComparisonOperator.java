package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

public enum CollectionComparisonOperator {
  ANY,
  ALL,
  NONE
}
