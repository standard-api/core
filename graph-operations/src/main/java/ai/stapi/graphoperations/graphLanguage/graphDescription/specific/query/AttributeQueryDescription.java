package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractAttributeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AttributeDescriptionParameters;
import java.util.ArrayList;
import java.util.List;

public class AttributeQueryDescription extends AbstractAttributeDescription implements QueryGraphDescription {

  public static final String SERIALIZATION_TYPE = "AttributeQueryDescription";
  public static final String DESCRIBED_STRUCTURE_TYPE = "Irrelevant";

  private AttributeQueryDescription() {
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      GraphDescription... childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, new ArrayList<>());
  }
}
