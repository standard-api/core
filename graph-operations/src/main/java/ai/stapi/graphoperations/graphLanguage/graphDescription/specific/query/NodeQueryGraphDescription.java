package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractNodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import java.util.List;

public class NodeQueryGraphDescription extends AbstractNodeDescription
    implements GraphElementQueryDescription {

  public static final String SERIALIZATION_TYPE = "72ba0bef-5f38-489b-bbbf-8212191f1f92";

  private SearchQueryParameters searchQueryParameters;

  protected NodeQueryGraphDescription() {
    super();
  }

  public NodeQueryGraphDescription(NodeDescription nodeDescription) {
    super(
        SERIALIZATION_TYPE,
        (NodeDescriptionParameters) nodeDescription.getParameters(),
        nodeDescription.getChildGraphDescriptions()
    );
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      List<GraphDescription> childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.parameters = parameters;
    this.searchQueryParameters = searchQueryParameters;
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.parameters = parameters;
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters
  ) {
    super(SERIALIZATION_TYPE, parameters);
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
  }

  @Override
  public SearchQueryParameters getSearchQueryParameters() {
    return this.searchQueryParameters;
  }
}
