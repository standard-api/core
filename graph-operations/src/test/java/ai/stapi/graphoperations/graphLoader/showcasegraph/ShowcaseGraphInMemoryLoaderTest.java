package ai.stapi.graphoperations.graphLoader.showcasegraph;

import ai.stapi.graphoperations.fixtures.ShowcaseGraphFixturesProvider;
import ai.stapi.graphoperations.fixtures.testsystem.TestSystemModelDefinitionsLoader;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.graph.repositorypruner.RepositoryPruner;
import ai.stapi.graphoperations.synchronization.GraphSynchronizer;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AllAttributesDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.UuidIdentityDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.AttributeQueryDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.CollectionComparisonOperator;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.NodeQueryGraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.OutgoingEdgeQueryDescription;
import ai.stapi.graphoperations.graphLoader.GraphLoaderReturnType;
import ai.stapi.graphoperations.graphLoader.inmemory.InMemoryGraphLoader;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import ai.stapi.graphoperations.graphLoader.search.filterOption.AllMatchFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.AnyMatchFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.ContainsFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.EndsWithFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.EqualsFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.NoneMatchFilterOption;
import ai.stapi.graphoperations.graphLoader.search.paginationOption.OffsetPaginationOption;
import ai.stapi.graphoperations.graphLoader.search.sortOption.AscendingSortOption;
import ai.stapi.graphoperations.graphLoader.search.sortOption.DescendingSortOption;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(TestSystemModelDefinitionsLoader.SCOPE)
class ShowcaseGraphInMemoryLoaderTest extends SchemaIntegrationTestCase {

  @Autowired
  private InMemoryGraphLoader inMemoryGraphLoader;

  @BeforeAll
  @Order(2)
  public static void beforeAll(
      @Autowired GraphSynchronizer graphSynchronizer,
      @Autowired RepositoryPruner repositoryPruner,
      @Autowired ShowcaseGraphFixturesProvider showcaseGraphFixturesProvider
  ) {
    repositoryPruner.prune();
    graphSynchronizer.synchronize(showcaseGraphFixturesProvider.getFixtureGraph());
  }

  @Test
  void itShouldFilterListByAllFilterOption() {
    var allEquals = new AllMatchFilterOption(
        new EqualsFilterOption<>(
            "line",
            "28. rijna 50"
        )
    );
    var searchParam = SearchQueryParameters.from(allEquals);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters("Address"),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterListByAnyFilterOption() {
    var anyEndsWith = new AnyMatchFilterOption(
        new EndsWithFilterOption(
            "line",
            "1"
        )
    );
    var searchParam = SearchQueryParameters.from(anyEndsWith);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters("Address"),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterListByNoneFilterOption() {
    var noneContains = new NoneMatchFilterOption(
        new ContainsFilterOption(
            "line",
            "8"
        )
    );
    var searchParam = SearchQueryParameters.from(noneContains);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters("Address"),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenGraphElementsApproved(actual);
  }

  @Test
  void itShouldJoinComplexSortJoinedNodesAndMapIntoCompactObject() {
    var graphDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("StructureDefinition"),
        SearchQueryParameters.from(),
        new UuidIdentityDescription(),
        new AttributeQueryDescription("kind"),
        new AttributeQueryDescription("abstract"),
        new OutgoingEdgeQueryDescription(
            new EdgeDescriptionParameters("differential"),
            new SearchQueryParameters(),
            new NodeQueryGraphDescription(
                new NodeDescriptionParameters("StructureDefinitionDifferential"),
                new OutgoingEdgeQueryDescription(
                    new EdgeDescriptionParameters("element"),
                    SearchQueryParameters.builder()
                        .setPaginationOption(new OffsetPaginationOption(0, 100))
                        .addSortOption(
                            new AscendingSortOption(
                                new NodeDescription(
                                    new NodeDescriptionParameters("ElementDefinition"),
                                    new AttributeQueryDescription("path")
                                )
                            )
                        ).build(),
                    new NodeDescription(
                        new NodeDescriptionParameters("ElementDefinition"),
                        new UuidIdentityDescription(),
                        new AttributeQueryDescription("path"),
                        new AttributeQueryDescription("min"),
                        new AttributeQueryDescription("max")
                    )
                )
            )
        )
    );

    var actual = this.inMemoryGraphLoader.get(
        new UniqueIdentifier("CodeableConcept"),
        graphDescription,
        Object.class,
        GraphLoaderReturnType.GRAPH,
        GraphLoaderReturnType.OBJECT
    );

    this.thenObjectApproved(actual.getData());
  }

  @Test
  void itShouldJoinAndFilterDeeplyByListAttribute() {
    var filter = new AllMatchFilterOption(
        new ContainsFilterOption(
            new OutgoingEdgeQueryDescription(
                new EdgeDescriptionParameters("address"),
                SearchQueryParameters.from(),
                CollectionComparisonOperator.ANY,
                new NodeDescription(
                    new NodeDescriptionParameters("Address"),
                    new AttributeQueryDescription("line")
                )
            ),
            "c"
        )
    );
    var searchQueryParameters = SearchQueryParameters.builder()
        .addFilterOption(filter)
        .build();
    var actual = this.inMemoryGraphLoader.find(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters("Organization"),
            searchQueryParameters,
            new AttributeQueryDescription("name"),
            new OutgoingEdgeQueryDescription(
                new EdgeDescriptionParameters("address"),
                SearchQueryParameters.builder()
                    .addSortOption(
                        new AscendingSortOption(
                            new NodeDescription(
                                new NodeDescriptionParameters("Address"),
                                new AttributeQueryDescription("line")
                            )
                        )
                    ).build(),
                new NodeDescription(
                    new NodeDescriptionParameters("Address"),
                    new AttributeQueryDescription("country"),
                    new AttributeQueryDescription("line")
                )
            )
        ),
        Object.class,
        GraphLoaderReturnType.OBJECT
    );
    this.thenObjectApproved(actual.getData());
  }

  @Test
  void itShouldJoinDeeperAndSortSecondNodeWithDeepSort() {
    var innerSort = new AscendingSortOption(
        new NodeDescription(
            new NodeDescriptionParameters("ElementDefinitionType"),
            new AttributeQueryDescription("code")
        )
    );
    var sort = new DescendingSortOption(
        new NodeDescription(
            new NodeDescriptionParameters("ElementDefinition"),
            new OutgoingEdgeQueryDescription(
                new EdgeDescriptionParameters("type"),
                SearchQueryParameters.from(innerSort),
                new NodeQueryGraphDescription(
                    new NodeDescriptionParameters("ElementDefinitionType"),
                    SearchQueryParameters.from(),
                    new AttributeQueryDescription("code")
                )
            )
        )
    );
    var secondSort = new DescendingSortOption(
        new NodeDescription(
            new NodeDescriptionParameters("ElementDefinition"),
            new AttributeQueryDescription("path")
        )
    );
    var searchQueryParameters = SearchQueryParameters.builder()
        .addSortOption(sort)
        .addSortOption(secondSort)
        .build();

    var actual = this.inMemoryGraphLoader.find(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters("StructureDefinition"),
            SearchQueryParameters.builder()
                .addSortOption(new AscendingSortOption("type"))
                .setPaginationOption(new OffsetPaginationOption(0, 20))
                .build(),
            new AttributeQueryDescription("type"),
            new OutgoingEdgeQueryDescription(
                new EdgeDescriptionParameters("differential"),
                SearchQueryParameters.from(),
                new NodeQueryGraphDescription(
                    new NodeDescriptionParameters("StructureDefinitionDifferential"),
                    SearchQueryParameters.from(),
                    new OutgoingEdgeQueryDescription(
                        new EdgeDescriptionParameters("element"),
                        searchQueryParameters,
                        new NodeDescription(
                            new NodeDescriptionParameters("ElementDefinition"),
                            new AttributeQueryDescription("path"),
                            new AttributeQueryDescription("min"),
                            new AttributeQueryDescription("max"),
                            new OutgoingEdgeQueryDescription(
                                new EdgeDescriptionParameters("type"),
                                SearchQueryParameters.from(innerSort),
                                new NodeDescription(
                                    new NodeDescriptionParameters("ElementDefinitionType"),
                                    new AttributeQueryDescription("code")
                                )
                            )
                        )
                    )
                )
            )
        ),
        Object.class,
        GraphLoaderReturnType.OBJECT
    );
    this.thenObjectApproved(actual.getData());
  }
}
