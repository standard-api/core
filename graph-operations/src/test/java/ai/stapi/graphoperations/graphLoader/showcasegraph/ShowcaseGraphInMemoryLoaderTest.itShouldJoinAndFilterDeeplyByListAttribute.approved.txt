[ {
  "name" : "Prague Medical Clinic",
  "address" : [ {
    "country" : "Czech Republic",
    "line" : [ "Dlouh√° 21" ]
  }, {
    "country" : "Czech Republic",
    "line" : [ "Vaclavske Namesti 8", "Vaclavske Namesti 11", "Vaclavske Namesti 50" ]
  } ]
} ]